railties/lib/rails/application.rb

```diff
@@ -247,9 +247,11 @@ def config_for(name, env: Rails.env)
         config, shared = all_configs[env.to_sym], all_configs[:shared]
 
         if shared
-          config = {} if config.nil?
-          if config.is_a?(Hash)
+          config = {} if config.nil? && shared.is_a?(Hash)
+          if config.is_a?(Hash) && shared.is_a?(Hash)
             config = shared.deep_merge(config)
+          elsif config.nil? && shared.is_a?(Array)
```

Comment:
I don't think you need the `shared.is_a?(Array)`. If `config` is `nil`, then we get the entire `shared` regardless of it's type.

Reply:
Indeed